ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,489 @@
+-- as objectors the same chain as the dynamo.
+--
+onRollBackward :: forall m peer blk.
+  ( MonadSTM m,
+    Eq peer,
+    HasHeader (Header blk),
+    NoThunks (Header blk)
+  ) =>
+  PeerContext m peer blk ->
+  WithOrigin SlotNo ->
+  STM m ()
+onRollBackward context slot = whenEnabled context () $
+  readTVar (cschJumping (handle context)) >>= \case
+    Objector badPoint
+      | slot < pointSlot badPoint -> do
+          disengage (handle context)
+          electNewObjector (stripContext context)
+      | otherwise -> pure ()
+    Disengaged -> pure ()
+    Jumper{} -> pure ()
+    Dynamo lastJumpSlot
+      | slot < lastJumpSlot -> do
+          disengage (handle context)
+          electNewDynamo (stripContext context)
+      | otherwise -> pure ()
+
+-- | This function is called when we receive a 'MsgAwaitReply' message.
+--
+-- If this is the dynamo, we need to elect a new dynamo as no more headers
+-- are available.
+onAwaitReply ::
+  ( MonadSTM m,
+    Eq peer,
+    HasHeader (Header blk),
+    NoThunks (Header blk)
+  ) =>
+  PeerContext m peer blk ->
+  STM m ()
+onAwaitReply context = whenEnabled context () $
+  readTVar (cschJumping (handle context)) >>= \case
+    Dynamo{} -> do
+      disengage (handle context)
+      electNewDynamo (stripContext context)
+    Objector{} -> do
+      disengage (handle context)
+      electNewObjector (stripContext context)
+    Jumper{} ->
+      -- A jumper might be receiving a 'MsgAwaitReply' message if it was
+      -- previously an objector and a new dynamo was elected.
+      disengage (handle context)
+    Disengaged ->
+      pure ()
+
+-- | Process the result of a jump. In the happy case, this only consists in
+-- updating the peer's handle to take the new candidate fragment and the new
+-- last jump point into account. When disagreeing with the dynamo, though, we
+-- enter a phase of several jumps to pinpoint exactly where the disagreement
+-- occurs. Once this phase is finished, we trigger the election of a new
+-- objector, which might update many TVars.
+processJumpResult :: forall m peer blk.
+  ( MonadSTM m,
+    HasHeader blk,
+    HasHeader (Header blk),
+    NoThunks (Header blk)
+  ) =>
+  PeerContext m peer blk ->
+  JumpResult blk ->
+  STM m ()
+processJumpResult context jumpResult = whenEnabled context () $
+  readTVar (cschJumping (handle context)) >>= \case
+    Dynamo _ -> pure ()
+    Disengaged -> pure ()
+    Objector _ -> pure ()
+    Jumper nextJumpVar goodPoint jumperState ->
+        case jumpResult of
+          AcceptedJump fragment -> do
+            -- The jump was accepted; we set the jumper's candidate fragment to
+            -- the dynamo's candidate fragment up to the accepted point.
+            --
+            -- The candidate fragments of jumpers don't grow otherwise, as only the
+            -- objector and the dynamo request further headers.
+            modifyTVar (cschState (handle context)) $ \csState -> csState {csCandidate = fragment}
```

Comment:
Is there any restriction to set the candidate fragment here? e.g. does it need to be anchored at the immutable tip or does it need to contain the immutable tip?